docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+ * @param[out] params The evidence parameters.
+ * @param[out] params_size The size of the evidence parameter.
+ * @retval OE_OK on success.
+ * @retval OE_NOT_FOUND if none of the format ids match any registered verifier.
+ * @retval other appropriate error code.
+ */
+oe_result_t oe_get_evidence_parameters(
+    const oe_uuid_t* format_id,
+    uint8_t** params,
+    size_t* params_size);
+
+/**
+ * oe_free_evidence_parameters
+ *
+ * Frees the evidence parameters.
+ *
+ * @param[in] params The evidence parameters.
+ * @retval OE_OK on success.
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
+ * @retval other appropriate error code.
+ */
+ oe_result_t oe_free_evidence_parameters(uint8_t* params);
+```
+### Deprecate built-in SGX plugin and local/remote flags
+```C
+#if (OE_API_VERSION < 3)
+/**
+ * Helper function that returns the SGX attester that can then be sent to
+ * `oe_register_attester`.
+ * 
+ * @deprecated This function is deprecated. Use oe_sgx_get_attesters instead.
+ *
+ * @retval A pointer to the SGX attester. This function never fails.
+ */
+oe_attester_t* oe_sgx_plugin_attester(void);
+
+/**
+ * Helper function that returns the SGX verifier that can then be sent to
+ * `oe_register_verifier`.
+ *
+ * @deprecated This function is deprecated. Use oe_sgx_get_verifiers instead.
+ * 
+ * @retval A pointer to the SGX verifier. This function never fails.
+ */
+oe_verifier_t* oe_sgx_plugin_verifier(void);
+
+/**
+ * Flags passed to oe_get_evidence() function.
+ * 
+ * @deprecated These flags are deprecated. Specify attester format
+ * IDs that are specifically for local or remote attestations instead.
+ * 
+ */
+#define OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION 0x00000000
+#define OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION 0x00000001
+#endif
+```
+### New verifier callback function
+
+```C
+/**
+ * The verifier attestion role. The verifier is reponsible for verifying the
+ * attestation evidence and must implement the functions below.
+ */
+typedef struct _oe_verifier oe_verifier_t;
+struct _oe_verifier
+{
+    ...
+
+    /**
+     * get_evidence_parameters
+     *
+     * Get evidence parameters that are to be used by attester as optional
+     * inputs when generating evidence.
+     *
+     * @param[out] params The evidence parameters.
+     * @param[out] params_size The size of the evidence parameter.
+     * @retval OE_OK on success.
+     * @retval other appropriate error code.
+     */
+    oe_result_t (*get_evidence_parameters(
+        uint8_t** params,
+        size_t* params_size);
+};
+```
+
+## Code samples
+More plugin code snippets are availabe in [CustomAttestation.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md). The sample below is mainly for demonstrating the use of new APIs.
+
+The plugin-ware verifier application, which can either be the enclave or the host, can verify the evidence like this:
```

Comment:
Should "plugin-ware" be "plugin-aware"?

Reply:
Fixed.

---
In reply to: [393257507](https://github.com/openenclave/openenclave/pull/2621#discussion_r393257507) [](ancestors = 393257507)